resource-manager/src/main/java/org/snomed/module/storage/ModuleStorageCoordinator.java

```diff
@@ -0,0 +1,602 @@
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String asMetadataResourcePath(String packageResourcePath) {
+        String[] splits = packageResourcePath.split("/");
+        splits = Arrays.copyOf(splits, splits.length - 1); // Remove last segment, i.e. a/b/c/d => a/b/c
+
+        return String.join("/", splits) + "/metadata.json";
+    }
+
+    private ModuleMetadata doDownloadRelease(String codeSystem, String moduleId, String effectiveTime, boolean includeFile) {
+        for (String readDirectory : this.readDirectories) {
+            LOGGER.trace("Downloading {}/{}_{}/{}", readDirectory, codeSystem, moduleId, effectiveTime);
+            String baseResourcePath = getBaseResourcePath(readDirectory, codeSystem, moduleId, effectiveTime);
+            String metadataResourcePath = getMetadataResourcePath(readDirectory, codeSystem, moduleId, effectiveTime);
+            if (resourceManager.doDoesObjectExist(metadataResourcePath)) {
+                ModuleMetadata moduleMetadata = FileUtils.convertToObject(resourceManager.doReadResourceFile(metadataResourcePath), ModuleMetadata.class);
+                if (moduleMetadata != null) {
+                    String rf2ResourcePatch = baseResourcePath + "/" + moduleMetadata.getFilename();
+                    if (resourceManager.doDoesObjectExist(rf2ResourcePatch)) {
+                        if (includeFile) {
+                            moduleMetadata.setFile(resourceManager.doReadResourceFile(rf2ResourcePatch));
+                        }
+
+                        return moduleMetadata;
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private boolean doArchiveRelease(String codeSystem, String moduleId, String effectiveTime) {
+        LOGGER.trace("Archiving {}/{}_{}/{}", writeDirectory, codeSystem, moduleId, effectiveTime);
+
+        String metadataResourcePath = getMetadataResourcePath(writeDirectory, codeSystem, moduleId, effectiveTime);
+        if (!resourceManager.doDoesObjectExist(metadataResourcePath)) {
+            return false;
+        }
+
+        ModuleMetadata moduleMetadata = FileUtils.convertToObject(resourceManager.doReadResourceFile(metadataResourcePath), ModuleMetadata.class);
+        if (moduleMetadata == null) { // De-serialisation failed
+            return false;
+        }
+
+        String packageResourcePath = getPackageResourcePath(writeDirectory, codeSystem, moduleId, effectiveTime, moduleMetadata.getFilename());
+        if (!resourceManager.doDoesObjectExist(packageResourcePath)) {
+            return false;
+        }
+
+        String epochSecond = Long.toString(Instant.now().getEpochSecond());
+        String metadataArchivePath = asArchivePath(metadataResourcePath, epochSecond);
+        String packageArchivePath = asArchivePath(packageResourcePath, epochSecond);
+        boolean metadataCopied = resourceManager.doCopyResource(metadataResourcePath, metadataArchivePath);
```

Comment:
@pgwilliams Archive could only work once because potentially the names would be taken (i.e. metadata.json cannot be archived because another metadata.json has been archived). I've therefore created a directory inside the archive folder being the epoch when it was archived. This allows for multiple same-named files to be deleted and also has an audit log.